File: leetcode/array/medium/golang_solutions/bfs_dfs/max_area_of_island_695.go
package bfs_dfs

func maxAreaOfIsland(grid [][]int) (max int) {
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == 1 {
				if currentMax := findMaxIsland(grid, i, j); currentMax > max {
					max = currentMax
				}
			}
		}
	}
	return
}

func findMaxIsland(grid [][]int, x, y int) int {
	if x < 0 || x >= len(grid) {
		return 0
	}
	if y < 0 || y >= len(grid[0]) {
		return 0
	}
	if grid[x][y] == 0 {
		return 0
	}
	grid[x][y] = 0
	return findMaxIsland(grid, x+1, y) +
		findMaxIsland(grid, x-1, y) +
		findMaxIsland(grid, x, y+1) +
		findMaxIsland(grid, x, y-1) + 1
}

File: leetcode/array/medium/java_solutions/MaxAreaOfIsland_695.java
package leetcode.array.medium.java_solutions;

public class MaxAreaOfIsland_695 {

    public int maxAreaOfIsland(int[][] grid) {
        var max = 0;
        for (var i = 0; i < grid.length; i++) {
            for (var j = 0; j < grid[0].length; j++) {
                var currentMax = exploreIsland(grid, i, j);
                if (max < currentMax) {
                    max = currentMax;
                }
            }
        }
        return max;
    }

    private int exploreIsland(int[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length) {
            return 0;
        }
        if (grid[i][j] == 0) {
            return 0;
        }
        grid[i][j] = 0;
        return exploreIsland(grid, i + 1, j) +
                exploreIsland(grid, i - 1, j) +
                exploreIsland(grid, i, j + 1) +
                exploreIsland(grid, i, j - 1) + 1;
    }

}
